plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,302 @@
+        self['boxType'] = 'plot'
+        self['fileId'] = fileId
+        self['shareKey'] = shareKey
+        self['title'] = title
+
+
+class Container(dict):
+    def __init__(self, box_1=EmptyBox(), box_2=EmptyBox(), size=400,
+                 sizeUnit='px', direction='vertical'):
+        self['type'] = 'split'
+        self['size'] = size
+        self['sizeUnit'] = sizeUnit
+        self['direction'] = direction
+        self['first'] = box_1
+        self['second'] = box_2
+
+
+class Dashboard(dict):
+    def __init__(self, dashboard_json=None, backgroundColor='#FFFFFF',
+                 boxBackgroundColor='#ffffff', boxBorderColor='#d8d8d8',
+                 boxHeaderBackgroundColor='#f8f8f8', foregroundColor='#333333',
+                 headerBackgroundColor='#2E3A46', headerForegroundColor='#FFFFFF',
+                 links=[], logoUrl='', title='Untitled Dashboard'):
+        self.box_ids_dict = {}
+        if not dashboard_json:
+            self['layout'] = EmptyBox()
+            self['version'] = 2
+            self['settings'] = {
+                'backgroundColor': backgroundColor,
+                'boxBackgroundColor': boxBackgroundColor,
+                'boxBorderColor': boxBorderColor,
+                'boxHeaderBackgroundColor': boxHeaderBackgroundColor,
+                'foregroundColor': foregroundColor,
+                'headerBackgroundColor': headerBackgroundColor,
+                'headerForegroundColor': headerForegroundColor,
+                'links': links,
+                'logoUrl': logoUrl,
+                'title': title
+            }
+            # TODO: change name to box_id_to_path
+        else:
+            self['layout'] = dashboard_json['layout']
+            self['version'] = dashboard_json['layout']
+            self['settings'] = dashboard_json['settings']
+
+            all_nodes = []
+            node_gen = node_generator(dashboard_json['layout'])
+
+            finished_iteration = False
+            while not finished_iteration:
+                try:
+                    all_nodes.append(node_gen.next())
+                except StopIteration:
+                    finished_iteration = True
+
+            for node in all_nodes:
+                if (node[1] != () and node[0]['type'] == 'box' and
+                    node[0]['boxType'] != 'empty'):
+                    try:
+                        max_id = max(self.box_ids_dict.keys())
+                    except ValueError:
+                        max_id = 0
+                    self.box_ids_dict[max_id + 1] = list(node[1])
+
+    def _insert(self, box_or_container, array_of_paths):
+        """Performs user-unfriendly box and container manipulations."""
+        if any(path not in ['first', 'second'] for path in array_of_paths):
+            raise exceptions.PlotlyError(
+                "Invalid path. Your 'array_of_paths' list must only contain "
+                "the strings 'first' and 'second'."
+            )
```

Comment:
🐄 this is a private function, do we really need to protect it from bad call arguments? if you feel like it's important to protect it, that's OK, but it may not be necessary. Just a thought :)